servicetalk-log4j2-mdc/src/main/java/io/servicetalk/log4j2/mdc/MdcCapturedContextProvider.java

```diff
@@ -0,0 +1,109 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.log4j2.mdc;
+
+import io.servicetalk.concurrent.api.CapturedContext;
+import io.servicetalk.concurrent.api.CapturedContextProvider;
+import io.servicetalk.concurrent.api.Scope;
+import io.servicetalk.context.api.ContextMap;
+
+import org.apache.logging.log4j.ThreadContext;
+import org.apache.logging.log4j.spi.ReadOnlyThreadContextMap;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * A {@link CapturedContextProvider} implementation to correctly propagate MDC context.
+ * Note: this class should not be used directly: it is only intended to be service loaded by ServiceTalk.
+ */
+public final class MdcCapturedContextProvider implements CapturedContextProvider {
+
+    private final boolean enabled;
+
+    /**
+     * Create a new {@link CapturedContextProvider} for MDC.
+     * Note: this class should not be used directly: it is only intended to be service loaded by ServiceTalk.
+     */
+    public MdcCapturedContextProvider() {
+        enabled = shouldEnableMdcCapture();
+    }
+
+    @Override
+    public CapturedContext captureContext(CapturedContext underlying) {
+        return enabled ? new MdcCapturedContext(underlying, getCurrent()) : underlying;
+    }
+
+    @Override
+    public CapturedContext captureContextCopy(CapturedContext underlying) {
+        return enabled ? new MdcCapturedContext(underlying, getCurrentCopy()) : underlying;
+    }
+
+    private static final class MdcCapturedContext implements CapturedContext {
+
+        private final CapturedContext delegate;
+        private final ConcurrentMap<String, String> storage;
+
+        MdcCapturedContext(CapturedContext delegate, ConcurrentMap<String, String> storage) {
+            this.delegate = delegate;
+            this.storage = storage;
+        }
+
+        @Override
+        public ContextMap captured() {
+            return delegate.captured();
+        }
+
+        @Override
+        public Scope attachContext() {
+            ConcurrentMap<String, String> old = getCurrent();
+            setCurrent(storage);
+            Scope delegateScope = delegate.attachContext();
+            return () -> {
+                delegateScope.close();
+                setCurrent(old);
+            };
+        }
+    }
+
+    private static ConcurrentMap<String, String> getCurrent() {
+        return DefaultServiceTalkThreadContextMap.CONTEXT_STORAGE.get();
+    }
+
+    private static ConcurrentMap<String, String> getCurrentCopy() {
+        ConcurrentMap<String, String> current = DefaultServiceTalkThreadContextMap.CONTEXT_STORAGE.get();
+        ConcurrentMap<String, String> result = new ConcurrentHashMap<>(Math.max(current.size(), 4));
```

Comment:
The default load-factor is 0.75, should we increase the size to accommodate `current.size()` values without extra resizing?

Reply:
I multiplied the current size by 1.5 to make the ending load factor a bit less than 0.75. I don't know if we have different prior art for this that suggests a different multiple.